OpenSim/Common/Test/testC3DFileAdapter.cpp

```diff
@@ -31,75 +31,163 @@
 #include <thread>
 #include <cmath>
 
-template<typename> class shrik;
-
-void compare_tables(const OpenSim::TimeSeriesTableVec3& table1,
-                    const OpenSim::TimeSeriesTableVec3& table2) {
+template<typename ETY = SimTK::Real>
+void compare_tables(const OpenSim::TimeSeriesTable_<ETY>& table1,
+            const OpenSim::TimeSeriesTable_<ETY>& table2,
+            const double tolerance = SimTK::SignificantReal) {
     using namespace OpenSim;
     try {
         OPENSIM_THROW_IF(table1.getColumnLabels() != table2.getColumnLabels(),
                          Exception,
                          "Column labels are not the same for tables.");
-        OPENSIM_THROW_IF(table1.getIndependentColumn() != 
-                         table2.getIndependentColumn(),
-                         Exception,
-                         "Independent columns are not the same for tables.");
+
+        ASSERT_EQUAL( table1.getIndependentColumn(), 
+                      table2.getIndependentColumn(), tolerance,
+                       __FILE__, __LINE__,
+                         "Independent columns are not equivalent.");
     } catch (const OpenSim::KeyNotFound&) {}
 
     const auto& matrix1 = table1.getMatrix();
     const auto& matrix2 = table2.getMatrix();
+
     for(int r = 0; r < matrix1.nrow(); ++r)
         for(int c = 0; c < matrix1.ncol(); ++c) {
             auto elt1 = matrix1.getElt(r, c); 
             auto elt2 = matrix2.getElt(r, c);
-            ASSERT_EQUAL(elt1[0], elt2[0], 1e-6);
-            ASSERT_EQUAL(elt1[1], elt2[1], 1e-6);
-            ASSERT_EQUAL(elt1[2], elt2[2], 1e-6);
+
+            ASSERT_EQUAL(elt1, elt2, tolerance, __FILE__, __LINE__,
+                "Element at row, " + std::to_string(r) + " col, " +
+                std::to_string(c) + " failed to have matching value.");
         }
 }
 
+template<typename ETY = SimTK::Real>
+void downsample_table(OpenSim::TimeSeriesTable_<ETY>& table,
+    const unsigned int increment) {
+    for (size_t r = table.getNumRows() - 2; r > 0; --r) {
+        if (r%increment)
+            table.removeRowAtIndex(r);
```

Comment:
This looks like it might be inefficient. Does the table resize/reorganize in memory when rows are removed? I expected a new table to be created. Also, because you're stepping through backwards, it seems like you could end up with strange gaps at the beginning (e.g., rows 1, 2, 5, 8, 11, ...).

Reply:
A table with 10 rows will end up with rows 0, 1, 2, 4, 5, 7, 8, 9, which is an unexpected result after "downsampling" by 3.

Reply:
> This looks like it might be inefficient. Does the table resize/reorganize in memory when rows are removed? I expected a new table to be created.

Efficiency isn't of any real concern here. Down sampling added no significant time to the test, and the downstream time savings are much more significant. In any case, removing in place does not reallocate.

> A table with 10 rows will end up with rows 0, 1, 2, 4, 5, 7, 8, 9, which is an unexpected result after "downsampling" by 3. 

It removes whenever their is a remainder, so in this case t looks like [0, 3, 6, 9] when the increment is 3.

Reply:
Oh ok, my bad. Thanks for clarifying.

Reply:
I still think this algorithm produces an unexpected result when the number of rows in the table is not 1 greater than a multiple of `increment`.

| Number of rows in original table | Rows retained when `increment` is 3 | Expected |
| --- | --- | --- |
| 4 | 0, 3 | 0, 3 |
| 5 | 0, 3, 4 | 0, 3 |
| 6 | 0, 3, 5 | 0, 3 |
| 7 | 0, 3, 6 | 0, 3, 6 |

Reply:
I intentionally left the start and end rows. This method is not available or intended for public use. It is just a quick way to speed up the test and reduce repo bloat with standard data files. It may be useful to have a down-sampling utility for users, but this is not it. That's why it only exists in this test.